### PR TITLE
[clang][Interp] Only diagnose null field access in constant contexts

### PR DESCRIPTION
**Pull request by tbaederr**
Looks like this should work as long as we don't dereference the value.

Note: This is a change from my local branch where I introduce integral pointers. One thing I changed is that null pointers are integral pointers and not block pointers, but in this version they are still block pointers so we have to be careful about `Pointee` being null.

### CODE REVIEW

**Comment by tbaederr** [comment]
Ping

File: clang/lib/AST/Interp/Pointer.h [hunk 1]
```diff
@@ -295,7 +295,7 @@ class Pointer {
   bool isUnion() const;
 
   /// Checks if the storage is extern.
-  bool isExtern() const { return Pointee->isExtern(); }
+  bool isExtern() const { return Pointee && Pointee->isExtern(); }
```
Comment (AaronBallman) [inline clang/lib/AST/Interp/Pointer.h:298]: If we need to care that `Pointee` can be null, we need to care consistently, don't we?

Comment (tbaederr) [reply clang/lib/AST/Interp/Pointer.h:298]: Kinda, we already care in several places, but I think when interpreting we end up checking `isZero()` or `isLive()` before calling any other `Pointer` API.

Comment (tbaederr) [reply clang/lib/AST/Interp/Pointer.h:298]: We could. Do you want me to add them?

Comment (AaronBallman) [reply clang/lib/AST/Interp/Pointer.h:298]: Yeah, I think so, just to help catch any uses that are out-of-contract.

**Review by AaronBallman — APPROVED** [review]
LGTM

**Pull request merged**